src/qdmi/Driver.cpp

```diff
@@ -0,0 +1,479 @@
+        dlsym(libHandle_, symbolName.c_str()));                                \
+    if ((symbol) == nullptr) {                                                 \
+      throw std::runtime_error("Failed to load symbol: " + symbolName);        \
+    }                                                                          \
+  }
+  try {
+    // NOLINTBEGIN(cppcoreguidelines-pro-type-reinterpret-cast)
+    // load the function symbols from the dynamic library
+    LOAD_DYNAMIC_SYMBOL(device_initialize)
+    LOAD_DYNAMIC_SYMBOL(device_finalize)
+    // device session interface
+    LOAD_DYNAMIC_SYMBOL(device_session_alloc)
+    LOAD_DYNAMIC_SYMBOL(device_session_init)
+    LOAD_DYNAMIC_SYMBOL(device_session_free)
+    LOAD_DYNAMIC_SYMBOL(device_session_set_parameter)
+    // device job interface
+    LOAD_DYNAMIC_SYMBOL(device_session_create_device_job)
+    LOAD_DYNAMIC_SYMBOL(device_job_free)
+    LOAD_DYNAMIC_SYMBOL(device_job_set_parameter)
+    LOAD_DYNAMIC_SYMBOL(device_job_query_property)
+    LOAD_DYNAMIC_SYMBOL(device_job_submit)
+    LOAD_DYNAMIC_SYMBOL(device_job_cancel)
+    LOAD_DYNAMIC_SYMBOL(device_job_check)
+    LOAD_DYNAMIC_SYMBOL(device_job_wait)
+    LOAD_DYNAMIC_SYMBOL(device_job_get_results)
+    // device query interface
+    LOAD_DYNAMIC_SYMBOL(device_session_query_device_property)
+    LOAD_DYNAMIC_SYMBOL(device_session_query_site_property)
+    LOAD_DYNAMIC_SYMBOL(device_session_query_operation_property)
+    // NOLINTEND(cppcoreguidelines-pro-type-reinterpret-cast)
+  } catch (const std::exception&) {
+    dlclose(libHandle_);
+    throw;
+  }
+  // initialize the device
+  device_initialize();
+}
+
+DynamicDeviceLibrary::~DynamicDeviceLibrary() {
+  // Check if QDMI_device_finalize is not NULL before calling it.
+  if (device_finalize != nullptr) {
+    device_finalize();
+  }
+  // close the dynamic library
+  if (libHandle_ != nullptr) {
+    dlclose(libHandle_);
+  }
+}
+} // namespace qdmi
+#endif // _WIN32
+
+QDMI_Device_impl_d::QDMI_Device_impl_d(
+    std::unique_ptr<qdmi::DeviceLibrary>&& lib)
+    : library_(std::move(lib)) {
+  if (library_->device_session_alloc(&deviceSession_) != QDMI_SUCCESS) {
+    throw std::runtime_error("Failed to allocate device session");
+  }
+  if (library_->device_session_init(deviceSession_) != QDMI_SUCCESS) {
+    library_->device_session_free(deviceSession_);
+    throw std::runtime_error("Failed to initialize device session");
+  }
+}
+
+auto QDMI_Device_impl_d::createJob(QDMI_Job* job) -> int {
+  if (job == nullptr) {
+    return QDMI_ERROR_INVALIDARGUMENT;
+  }
+  QDMI_Device_Job deviceJob = nullptr;
+  auto result =
+      library_->device_session_create_device_job(deviceSession_, &deviceJob);
+  if (result != QDMI_SUCCESS) {
+    return result;
+  }
+  auto uniqueJob =
+      std::make_unique<QDMI_Job_impl_d>(library_.get(), deviceJob, this);
+  *job = jobs_.emplace(uniqueJob.get(), std::move(uniqueJob)).first->first;
```

Comment:
I will keep commenting on that pattern until we find a more readable solution. this `emplace(...).first->first` is hard to grasp every time I look at it in a GitHub diff.